doc/crdt-details.md

```diff
@@ -0,0 +1,1086 @@
+    let new_inserts = union_ins_delta.inserted_subset();
+    if !new_inserts.is_empty() {
+        new_deletes = new_deletes.transform_expand(&new_inserts);
+    }
+
+    // 6. rebase insertions on text
+    let cur_deletes_from_union = &self.deletes_from_union;
+    let text_ins_delta = union_ins_delta.transform_shrink(cur_deletes_from_union);
+    // 7. apply inserts and rebase deletes_from_union
+    let text_with_inserts = text_ins_delta.apply(&self.text);
+    let rebased_deletes_from_union = cur_deletes_from_union.transform_expand(&new_inserts);
+
+    // 8. is the new edit in an undo group that was already undone due to concurrency?
+    let undone = self.undone_groups.contains(&undo_group);
+    let new_deletes_from_union = {
+        let to_delete = if undone { &new_inserts } else { &new_deletes };
+        rebased_deletes_from_union.union(to_delete)
+    };
+
+    // 9. move deleted or undone-inserted things from text to tombstones
+    let (new_text, new_tombstones) = shuffle(&text_with_inserts, &self.tombstones,
+        &rebased_deletes_from_union, &new_deletes_from_union);
+
+    let head_rev = &self.revs.last().unwrap();
+    (Revision {
+        rev_id: self.next_rev_id(),
+        max_undo_so_far: std::cmp::max(undo_group, head_rev.max_undo_so_far),
+        edit: Edit {
+            priority: new_priority,
+            undo_group: undo_group,
+            inserts: new_inserts,
+            deletes: new_deletes,
+        }
+    }, new_text, new_tombstones, new_deletes_from_union)
+}
+
+// TODO: have `base_rev` be an index so that it can be used maximally efficiently with the
+// head revision, a token or a revision ID. Efficiency loss of token is negligible but unfortunate.
+pub fn edit_rev(&mut self, priority: usize, undo_group: usize,
+        base_rev: RevToken, delta: Delta<RopeInfo>) {
+    let (new_rev, new_text, new_tombstones, new_deletes_from_union) =
+        self.mk_new_rev(priority, undo_group, base_rev, delta);
+    self.rev_id_counter += 1;
+    self.revs.push(new_rev);
+    self.text = new_text;
+    self.tombstones = new_tombstones;
+    self.deletes_from_union = new_deletes_from_union;
+}
+```
+
+#### InsertDelta::transform_expand
+
+This is another big function with lots of nested loops. Unlike `Subset::transform_expand` it takes a boolean parameter `after` that resolves the ambiguity of whether inserts should go before or after the characters being transformed over.
+
+```rust
+/// Do a coordinate transformation on an insert-only delta. The `after` parameter
+/// controls whether the insertions in `self` come after those specific in the
+/// coordinate transform.
+pub fn transform_expand(&self, xform: &Subset, after: bool) -> InsertDelta<N> {
+    // ...
+}
+```
+
+#### shuffle
+
+This helper is used in many different operations that use the pattern of computing how `deletes_from_union` should change and then using that to fix up the `text` and `tombstones` to correspond to that.
```

Comment:
should change and then ~using that to fix up the `text` and `tombstones` to correspond to that.~ **updating `text` and `tombstones` to reflect that new state.